# Jupyter Notebook

This is an example of a Jupyter Notebook.  Stephan introduced this last week; see the notes [here](https://github.com/raspstephan/ESS-Python-Tutorial/blob/master/materials/week1/jupyter-intro.ipynb).

__*You can shut a notebook down by closing browser windows, and from the terminal shell, typing ```ctrl``` + ```c``` twice.*__

# A brief introduction to Python

The official Python tutorial is very informative.  Most information below follows the official tutorial at https://docs.python.org/3/tutorial/.

In [1]:
2+2

4

In [2]:
70 - (5*6) / 4

62.5

In [3]:
(70 - (5*6)) / 4

10.0

## Floats versus integers

Python is a high-level language, and there are **no variable declarations** (Python does this implicitly).

* Numbers like 2 and 4 are seen as a type **```int```**, and numbers with decimals (62.5, 10.0) are seen as type **```float```**.
* **NOTE**: division **``` / ```** always returns a floating point number
* Get the type of a number or Python object using ```type()```

In [4]:
2/2

1.0

In [5]:
type(2/2)

float

In [6]:
2/1

2.0

## Floor division and modulo operators

Floor division **//** returns a division but discards the fractional remainder.

The modulo operator **%** returns only the remainder of the division.

In [7]:
7//4

1

In [8]:
7%4

3

## Powers in Python

To raise something to a power, use a double asterisk:  ** \*\* **

To ensure your numbers are interpreted as **```float```** types, good practice is to always use decimals in at least one number:

In [9]:
5**2

25

In [10]:
5**2.0

25.0

In [11]:
5.**2

25.0

## Assigning and printing

Assign variables using the ```=``` sign.

Print variables using the ```print()``` function.

In [12]:
width = 5.0
height = 7.0
area = width*height
print(area)

35.0


Note that Python supports very compact variable manipulation, allowing for ```+=```, ```-=```, ```*=```, and ```/=```

In [13]:
width = 5.0
width = width+5
print(width)

10.0


In [14]:
width = 5.0
width += 5
print(width)

10.0


In [15]:
width -= 5
print(width)

5.0


In [16]:
width *= 5
print(width)

25.0


In [17]:
width /= 5
print(width)

5.0


## Strings

Variables can also be assigned as strings when you enclose them in quotes.  __You can use _either_ single _or_ double quotes.__

In [18]:
string1 = "Hello"
print(string1)

Hello


In [19]:
string2 = 'w0rld'
print(string2)

w0rld


In [20]:
print(string1 + ', ' + string2)

Hello, w0rld


In [21]:
print(string1*3)

HelloHelloHello


Strings can also be indexed.  Remember the first index is ```0``` for Python.

In [22]:
string1[0]

'H'

In [23]:
string1[1]

'e'

You can also index backwards in Python.  The last character is at the index location ```-1```:

In [24]:
string1[-1]

'o'

In [25]:
string1[-3]

'l'

## Lists

This is a versatile way of creating compound data types.  Indexing works in a similar way as above.

Create a list by enclosing comma-separated items in square brackets:

In [26]:
list1 = [1,2,3,4,5,6,7,8,9]

In [27]:
print(list1)

[1, 2, 3, 4, 5, 6, 7, 8, 9]


In [28]:
print(list1[0])

1


In [29]:
print(list1[-3])

7


### Slicing lists

Slicing operations return new lists.

In [30]:
list1[3:]

[4, 5, 6, 7, 8, 9]

In [31]:
list2 = list1[3:]

In [32]:
print(list2)

[4, 5, 6, 7, 8, 9]


### Appending to lists

Add to the end of a list by using the ```append()``` method.

__Note ```list2``` is an object, and has methods that can be accessed via the ```.``` syntax__

In [33]:
list2.append(234)

In [34]:
print(list2)

[4, 5, 6, 7, 8, 9, 234]


### Assigning and replacing list values

You can also assign specific values, or do this to slices.

In [35]:
list3 = ['a', 'b', 'c', 'd', 'e', 'f', 'g']

In [36]:
print(list3)

['a', 'b', 'c', 'd', 'e', 'f', 'g']


In [37]:
list3[0] = 'A'

In [38]:
print(list3)

['A', 'b', 'c', 'd', 'e', 'f', 'g']


In [39]:
list3[0:3] = ['A','B','C']

In [40]:
print(list3)

['A', 'B', 'C', 'd', 'e', 'f', 'g']


### Getting the length of a list

Use the ```len()``` function to get the length of a list.

In [41]:
len(list1)

9

In [42]:
len(list2)

7

## Simple loops

Python's syntax does **NOT** require semicolons at the end of statements.

The Python language determines where loops begin and end **strictly from spacing**.  Use the same number of spaces or a **```tab```** for each loop.

Loops can begin with **```for```** and **```while```**.  Other statements, such as **```else```**, **```if```**, and **```elif```** also exist.  Note **```elif```** and **```else if```** are equivalent and both acceptable.

For example, to loop through weekday names and print each:

In [43]:
weekdays = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
for i in [0,1,2,3,4,5,6]:
    print(weekdays[i])

Monday
Tuesday
Wednesday
Thursday
Friday
Saturday
Sunday


### The ```range()```  and ```list()``` functions

**```range()```** is a function that creates an *internal* list (i.e., an [iterator])(https://wiki.python.org/moin/Iterator) of numbers over which you can iterate.

* ```range(N)``` will go FROM ```0``` TO ```N-1```

In [44]:
range(5) # goes from 0 to 4 inclusive

range(0, 5)

You can convert this into a Python list by using the **```list()```** function:

In [45]:
list(range(5))

[0, 1, 2, 3, 4]

The **```range()```** function takes at most 3 arguments:  start, stop, and interval.

In [46]:
list(range(2,5))

[2, 3, 4]

In [47]:
list(range(0,100,5)) # lets you go from 0 to 99 at intervals of 5

[0, 5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60, 65, 70, 75, 80, 85, 90, 95]

To add comments in Python code, note use of the ```#``` symbol, as above.

## For loops

* Python does not require **```end```** statements or semicolons.  Instead, the use of spacing in the loops implicitly tells python where loops begin and end.  Manually insert (4!) spaces (or use a **```tab```** on the keyboard) to struture a loop.

* All output in Python is suppressed automatically, unless you choose to **```print()```** it.  This means you don't need semicolons at the end of lines, and you'll get an error if you do.

* Note **```range()```** can be used in combination with the **```len()```** or other functions to make looping more streamlined:

In [48]:
for i in range(7):
    print(weekdays[i])

Monday
Tuesday
Wednesday
Thursday
Friday
Saturday
Sunday


In [49]:
for i in range(len(weekdays)):
    print(weekdays[i])

Monday
Tuesday
Wednesday
Thursday
Friday
Saturday
Sunday


__You can also simply loop over the list itself:__

In [50]:
for i in weekdays:
    print(i)

Monday
Tuesday
Wednesday
Thursday
Friday
Saturday
Sunday


All loops respond to spacing in the same way.  See below for an example of a **```while```** loop.

## Defining functions (an example with the Fibonacci series)

*image from [Wikipedia](https://en.wikipedia.org/wiki/Fibonacci_number)*

![Fibonacci series](https://upload.wikimedia.org/wikipedia/commons/thumb/b/bf/PascalTriangleFibanacci.svg/360px-PascalTriangleFibanacci.svg.png)

In [51]:
a,b = 0,1 # note multiple statements in one line
while b<10:
    print(b)
    a,b = b, a+b

1
1
2
3
5
8


In [52]:
def fib(n): # Write a Fibonacci series up to integer n
    a,b = 0,1
    while a<n:
        print(a)
        a,b = b, a+b
fib(1000)

0
1
1
2
3
5
8
13
21
34
55
89
144
233
377
610
987


More advanced information on defining functions can be found online at the tutorial link,  https://docs.python.org/3/tutorial/controlflow.html.

Information on all statements useful in loops can be found there, including
**```if```**, **```for```**, **```break```**, **```continue```**, **```pass```**, and others.

## The ```lambda``` keyword

Finally, the **```lambda```** keyword is very useful when defining functions.  __There's not enough time in this tutorial to go through all the uses of ```lambda```, but it's something you will likely come across when doing more advanced statistical function fitting.__

In short, the **```lambda```** keyword will help you create anonymous functions (i.e., functions not bound to a name.  Note the use of **```return```** here, as well.

In [53]:
def raise_to_power(n):
    """ This creates a polynomial that raises a number x to the nth power """
    return lambda x: x**n

In [54]:
f = raise_to_power(5)

In [55]:
f(0)

0

In [56]:
f(1)

1

In [57]:
f(2)

32

Also note the string after the **```def()```** statement above.  This is a *function annotation*, and anything enclosed in three quotes (**double** or **single**) will not be printed but is useful for multi-line documentation.

# Common data types in Python (and NumPy)

* tuple ```a=(1,2)```

* integer ```b=1```

* float ```c=1.``` or ```c=1.0```

* list ```d=[1,2,3,4]```

* dictionary ```dict = {'a':(1,2), 'b':1, 'c':1.0, 'd':[1,2,3,4] }```

In [58]:
dict = {'a':(1,2), 'b':1, 'c':1.0, 'd':[1,2,3,4] }

In [59]:
dict['b']

1

In [60]:
dict['d']

[1, 2, 3, 4]